pkg/loader/load_test.go

```diff
@@ -378,6 +378,141 @@ func (s *batchManagerSuite) TestShouldExecAccumulatedDMLs(c *check.C) {
 	c.Assert(bm.txns, check.HasLen, 1)
 }
 
+type txnManagerSuite struct{}
+
+var _ = check.Suite(&txnManagerSuite{})
+
+func (s *txnManagerSuite) TestRunTxnManager(c *check.C) {
+	input := make(chan *Txn)
+	dmls := []*DML{
+		{Tp: UpdateDMLType},
+		{Tp: InsertDMLType},
+		{Tp: UpdateDMLType},
+	}
+	txn := &Txn{DMLs: dmls}
+	txnManager := newTxnManager(14, input)
+	output := txnManager.run()
+	// send 5 txns (size 3) to txnManager, the 5th txn should get blocked at cond.Wait()
+	for i := 0; i < 5; i++ {
+		select {
+		case input <- txn:
+		case <-time.After(50 * time.Microsecond):
+			c.Fatal("txnManager gets blocked while receiving txns")
+		}
+	}
+	c.Assert(output, check.HasLen, 4)
+	// Next txn should be blocked
+	select {
+	case input <- txn:
+		c.Fatal("txnManager doesn't block the txn when room is not enough")
+	case <-time.After(50 * time.Microsecond):
+	}
+	c.Assert(output, check.HasLen, 4)
+	// pick one txn from output channel
+	select {
+	case t := <-output:
+		txnManager.pop(t)
+		c.Assert(t, check.DeepEquals, txn)
+	case <-time.After(50 * time.Microsecond):
+		c.Fatal("Fail to pick txn from txnManager")
+	}
+	// Now txn won't be blocked but txnManager should be blocked at cond.Wait()
+	select {
+	case input <- txn:
+	case <-time.After(50 * time.Microsecond):
+		c.Fatal("txnManager gets blocked while receiving txns")
+	}
+	// close txnManager and output should be closed when txnManager is closed
+	txnManager.Close()
+	outputClose := make(chan struct{})
+	go func() {
+		for t := range output {
+			c.Assert(t, check.DeepEquals, txn)
+		}
+		close(outputClose)
+	}()
+	select {
+	case <-outputClose:
+	case <-time.After(time.Second):
+		c.Fatal("txnManager fails to end run()... and close output channel in 1s, may get blocked")
+	}
+}
+
+func (s *txnManagerSuite) TestAddBigTxn(c *check.C) {
+	input := make(chan *Txn)
+	txnSmall := &Txn{DMLs: []*DML{{Tp: UpdateDMLType}}}
+	txnBig := &Txn{DMLs: []*DML{
+		{Tp: UpdateDMLType},
+		{Tp: InsertDMLType},
+		{Tp: UpdateDMLType},
+	}}
+	txnManager := newTxnManager(1, input)
+	output := txnManager.run()
+	select {
+	case input <- txnSmall:
+	case <-time.After(50 * time.Microsecond):
```

Comment:
Use `default`?

Reply:
In `txnManager.run()` it will start a new goroutine and start watching input channel which will take some time. Using `default` may get fault.